network/p2p/scoring/app_score_test.go

```diff
@@ -0,0 +1,295 @@
+
+	// all nodes subscribe to block topic (common topic among all roles)
+	_, err := con1Node.Subscribe(blockTopic, unittest.NetworkCodec(), unittest.AllowAllPeerFilter(), slashingViolationsConsumer)
+	require.NoError(t, err)
+
+	con2Sub, err := con2Node.Subscribe(blockTopic, unittest.NetworkCodec(), unittest.AllowAllPeerFilter(), slashingViolationsConsumer)
+	require.NoError(t, err)
+
+	// access node group
+	accessNodeSubs := make([]*pubsub.Subscription, len(accessNodeGroup))
+	for i, node := range accessNodeGroup {
+		accessNodeSubs[i], err = node.Subscribe(blockTopic, unittest.NetworkCodec(), unittest.AllowAllPeerFilter(), slashingViolationsConsumer)
+		require.NoError(t, err)
+	}
+
+	// let nodes reside on a full topology, hence no partition is caused by the topology.
+	p2pfixtures.LetNodesDiscoverEachOther(t, ctx, allNodes, allIds)
+
+	proposalMsg := p2pfixtures.MustEncodeEvent(t, unittest.ProposalFixture())
+	require.NoError(t, con1Node.Publish(ctx, blockTopic, proposalMsg))
+
+	// we check that within a one-second window the message is not received by the other honest consensus node.
+	// the one-second window is important because it triggers the heartbeat of the con1Node to perform a lazy pull (iHave).
+	// And con1Node may randomly choose con2Node as the peer to perform the lazy pull.
+	// However, when con2Node is not in the mesh of con1Node, it is deprived of the eager push from con1Node.
+	// Hence, con2Node will not receive the message within a one-second window.
+	ctx1s, cancel1s := context.WithTimeout(ctx, 1*time.Second)
+	defer cancel1s()
+	return !p2pfixtures.HasSubReceivedMessage(t, ctx1s, proposalMsg, con2Sub)
+}
+
+// TestFullGossipSubConnectivityAmongHonestNodesWithMaliciousMajority is part two of testing pushing access nodes to the edges of the network.
+// This test proves that if access nodes are PUSHED to the edge of the network, even their malicious majority cannot partition
+// the network of honest nodes.
+func TestFullGossipSubConnectivityAmongHonestNodesWithMaliciousMajority(t *testing.T) {
+	total := 10
+	for i := 0; i < total; i++ {
+		if !testFullGossipSubConnectivityAmongHonestNodesWithMaliciousMajority(t) {
+			// even one failure should not happen, as it means that malicious majority can partition the network
+			// with our peer scoring parameters.
+			require.Fail(t, "honest nodes could not exchange message on GossipSub")
+		}
+	}
+}
+
+// testFullGossipSubConnectivityAmongHonestNodesWithMaliciousMajority runs a GossipSub network with malicious majority of access nodes.
+// The malicious nodes mute GossipSub traffic from/to honest nodes.
+// The honest nodes push access nodes to the edge of the network.
+// The test returns true if two honest nodes CAN exchange GossipSub publish message despite malicious majority of access nodes.
+func testFullGossipSubConnectivityAmongHonestNodesWithMaliciousMajority(t *testing.T) bool {
```

Comment:
this and `testNetworkPartitionWithNoHonestPeerScoringInFullTopology` are the same except for the `con*Node` fixture right? Can you refactor this a bit so that there's only one test method? That way it's easier to verify the tests are identical except for the peer scoring method

Reply:
https://github.com/onflow/flow-go/pull/3379/commits/8b3a828511de2619dfa3a98412adfef79831c43e